compiler/noirc_evaluator/src/ssa/opt/array_get_from_if_else_result.rs

```diff
@@ -0,0 +1,248 @@
+use crate::ssa::{
+    ir::{function::Function, instruction::Instruction, types::Type, value::Value},
+    ssa_gen::Ssa,
+};
+
+impl Ssa {
+    // Given an original IfElse instruction is this:
+    //
+    //     v10 = if v0 then v2 else if v1 then v3
+    //
+    // and a later ArrayGet instruction is this:
+    //
+    //     v11 = array_get v4, index v4
+    //
+    // we optimize the latter to this:
+    //
+    //     v12 = array_get v2, index v4
+    //     v13 = array_get v3, index v4
+    //     v14 = if v0 then v12 else if v1 then v13
+    #[tracing::instrument(level = "trace", skip(self))]
+    pub(crate) fn array_get_from_if_else_result_optimization(mut self) -> Self {
+        for function in self.functions.values_mut() {
+            optimize_array_get_from_if_else_result(function);
+        }
+
+        self
+    }
+}
+
+fn optimize_array_get_from_if_else_result(function: &mut Function) {
+    let block = function.entry_block();
+    let dfg = &mut function.dfg;
+    let instructions = dfg[block].take_instructions();
+
+    for instruction_id in instructions {
+        // Only apply this optimization to ArrayGet
+        let Instruction::ArrayGet { array, index } = &dfg[instruction_id].clone() else {
+            dfg[block].instructions_mut().push(instruction_id);
+            continue;
+        };
+
+        // Don't optimize if the index is a constant (this is optimized later on in a different way)
+        if let Value::NumericConstant { .. } = &dfg[dfg.resolve(*index)] {
```

Comment:
Did you see any optimization differences without this check?

Reply:
When I originally coded this there was a performances regression and I though it might have been because of this. So I added this if... but it didn't change anything, and I left it (because it didn't change anything, but also because it also regressed and I didn't know how to continue working on this PR)

Reply:
Ah, okay. If there's still a regression without it we can keep it then. I was mostly curious what effect it had.